tests/test_transformations/test_knn_clustering.py

```diff
@@ -0,0 +1,205 @@
+        # AnnData object's obs attribute and if it correctly sets
+        # 'knn_features' in the AnnData object's uns attribute.
+        knn_clustering(
+            adata=self.syn_data,
+            features=self.syn_features,
+            annotation=self.annotation,
+            layer=self.layer,
+        )
+        self.assertIn("knn", self.syn_data.obs)
+        self.assertEqual(self.syn_data.uns["knn_features"], self.syn_features)
+
+    def test_output_annotation(self):
+        # This test checks if the function correctly adds the "output_layer"
+        # to the # AnnData object's obs attribute
+        output_annotation_name = "my_output_annotation"
+        knn_clustering(
+            adata=self.syn_data,
+            features=self.syn_features,
+            annotation=self.annotation,
+            layer=self.layer,
+            output_annotation=output_annotation_name,
+        )
+        self.assertIn(output_annotation_name, self.syn_data.obs)
+
+    def test_layer_none_case(self):
+        # This test checks if the function works correctly when layer is None.
+        knn_clustering(
+            adata=self.syn_data,
+            features=self.syn_features,
+            annotation=self.annotation,
+            layer=None,
+        )
+        self.assertIn("knn", self.syn_data.obs)
+        self.assertEqual(self.syn_data.uns["knn_features"], self.syn_features)
+
+    def test_invalid_k(self):
+        # This test checks if the function raises a ValueError when the
+        # k argument is not a positive integer.
+        with self.assertRaises(ValueError):
+            knn_clustering(
+                adata=self.syn_data,
+                features=self.syn_features,
+                annotation=self.annotation,
+                layer=self.layer,
+                k="invalid",
+            )
+
+    def test_trivial_label(self):
+        # This test checks if the data is fully labeled or missing labels for every datapoint
+
+        # all datapoints labeled
+        with self.assertRaises(ValueError):
+            knn_clustering(
+                adata=self.syn_data,
+                features=self.syn_features,
+                annotation="no_missing_classes",
+                layer=self.layer,
+            )
+
+        # no datapoints labeled
+        with self.assertRaises(ValueError):
+            knn_clustering(
+                adata=self.syn_data,
+                features=self.syn_features,
+                annotation="all_missing_classes",
+                layer=self.layer,
+            )
+
+    def test_clustering_accuracy(self):
+        knn_clustering(
+            adata=self.syn_data,
+            features=self.syn_features,
+            annotation=self.annotation,
+            layer="counts",
+            k=50,
+        )
+
+        self.assertIn("knn", self.syn_data.obs)
+        self.assertEqual(len(np.unique(self.syn_data.obs["knn"])), 2)
+
+    def test_associated_features(self):
```

Comment:
@oeldaghar, I believe that test is already covered part of _select_input_features.